Update and reimplement AddressLowering pass (for SIL opaque values).

atrick opened the pull request:
Merge the AddressLowering pass from its old development branch and update it so we can begin incrementally enabling it under a flag.

This has been reimplemented for simplicity. There's no point in looking at the old code.

Note: OSSA is required to correctly handle ownership when lowering. The original prototype attempted to "fake" OSSA for generic values, which wasn't viable. This new version has been completely updated for OSSA. The memory lifetime verifier is another prerequisite for this pass and is also now enabled.

The focus of the current implementation is SIL-level testing of the lowering algorithm. The objective of this algorithm is to avoid unnecessary copies by reusing storage when lowering composition of aggregates and merging subjects across phis. Tests have been developed for the algorithm's corner cases.

The next step is to test lowering immediately after SILGen and develop functional testing across the breadth of SILGen's output.

atrick committed:
Update and reimplement AddressLowering pass (for SIL opaque values).

Merge the AddressLowering pass from its old development branch and update
it so we can begin incrementally enabling it under a flag.

This has been reimplemented for simplicity. There's no point in
looking at the old code.

atrick committed:
Move -enable-sil-opaque-value to SILOptions.

atrick committed:
Add emitLoad/emitStore to OpaqueValue type lowering.

atrick committed:
SILModule::hasLoweredAddress

atrick commented:
@swift-ci test

atrick commented:
@eeckstein I rewrote the header comments for `ValueStorage`, added comments to `runOnFunction, and clarified SIL examples a bit.

atrick committed:
Add support for indirect tuple-type results.

This could happen as a result of specialization or concrete
address-only values.

For now, it's just tested by SIL unit tests.

eeckstein commented on lib/SILOptimizer/Mandatory/AddressLowering.h:
```diff
@@ -0,0 +1,282 @@
+
+  ValueStorage() { clear(); }
+
+  void clear() {
```
AFAICT, you are using the `clear` function only in the constructor. Why not just do a member wise initialization?

eeckstein commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -2,271 +2,535 @@
-  ValueStorage &getStorage(SILValue value) {
-    return valueVector[getOrdinal(value)].second;
-  }
+ValueStorage &ValueStorageMap::insertValue(SILValue value) {
```
`insertValue` should not return a reference to the key, but receive the mapped value(s) as a parameter: either a `ValueStorage` or just a SILValue (the `storageAddress`

atrick replied:
> insertValue should not return a reference to the key, but receive the mapped value(s) 
That makes sense now. It's was more flexible the old way, but no longer important.

eeckstein commented on lib/SILOptimizer/Mandatory/AddressLowering.h:
```diff
@@ -0,0 +1,282 @@
+  ValueStorage &insertValue(SILValue value);
+
+  /// Replace a value that is mapped to storage with another value. This allows
+  /// limited rewritting of original address-only values. For example, block
```
... rewriting ...

eeckstein commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -2,271 +2,535 @@
-      : F(F),
-        loweredFnConv(F->getLoweredFunctionType(),
-            SILModuleConventions::getLoweredAddressConventions(F->getModule())),
+  SmallVector<SILInstruction *, 8> exitingInsts;
```
Why not `SmallVector<TermInst *, 8>`?

eeckstein commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -282,312 +546,656 @@ class OpaqueValueVisitor {
-/// finding the lifetime boundary with a simple backward walk from uses.
-class OpaqueStorageAllocation {
-  AddressLoweringState &pass;
+// Canonicalize returned values.
```
Can you add a sentence here what this does? E.g. "Make sure that the operand of the return instruction is a tuple which with no other uses than the return instruction"

eeckstein reviewed (commented):
A few initial comments.
I'll continue reviewing next week

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -2,271 +2,521 @@
+/// reuse the borrowed value's storage. This means that we SIL cannot allow
+/// guaranteed opaque uses unless they are projections of the definition. In
+/// particular, borrowed structs, tuples, and enums of address-only types are
+/// not allowed.
```
SIL.rst
```
Forwarding Address-Only Values
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Address-only values are potentially unmovable when borrowed. This
means that they cannot be forwarded with guaranteed ownership unless
the forwarded value has the same representation as in the original
value and can reuse the same storage. Non-destructive projection is
allowed, such as `struct_extract`. Aggregation, such as `struct`, and
destructive disaggregation, such as `switch_enum` is not allowed. This
is an invariant for OSSA with opaque SIL values for these reasons:

1. To avoid implicit semantic copies. For move-only values, this allows
complete diagnostics. And in general, it makes it impossible for SIL
passes to "accidentally" create copies.

2. To reuse borrowed storage. This allows the optimizer to share the same
storage for multiple exclusive reads of the same variable, avoiding
copies. It may also be necessary to support native Swift atomics, which
will be unmovable-when-borrowed.
```
SILVerifier::checkOwnershipForwardingInst
```
    // Address-only values are potentially unmovable when borrowed. Ensure that
    // guaranteed address-only values are forwarded with the same
    // representation. Non-destructive projection is allowed. Aggregation and
    // destructive disaggregation is not allowed. See SIL.rst, Forwarding
    // Addres-Only Values.
    if (ownership == OwnershipKind::Guaranteed
        && OwnershipForwardingMixin::isAddressOnly(i)) {
      require(OwnershipForwardingMixin::hasSameRepresentation(i),
              "Forwarding a guaranteed address-only value requires the same "
              "representation since no move or copy is allowed.");
    }
```
SILVerifier::visitSILPhiArgument
```
        // Address-only values are potentially unmovable when borrowed. See also
        // checkOwnershipForwardingInst. A phi implies a move of its arguments
        // because they can't necessarilly all reuse the same storage.
        require((!arg->getType().isAddressOnly()
                 || arg->getOwnershipKind() != OwnershipKind::Guaranteed) &&
                "Guaranteed address-only phi not allowed--implies a copy");
```
AddressLowering file comment
```
/// This pass removes "opaque SILValues" by translating them into addressable
/// memory locations such as a stack locations. This is mandatory for IRGen.
///
/// Lowering to LLVM IR requires each SILValue's type to be a valid "SIL storage
/// type". Opaque SILValues have address-only types. Address-only values require
/// indirect storage in LLVM, so their SIL storage type must be an address type.
///
/// This pass never creates copies except to replace explicit value copies
/// (copy_value, load [copy], store). For move-only values, this allows complete
/// diagnostics. And in general, it makes it impossible for SIL passes to
/// "accidentally" create copies.
///
/// This pass inserts moves (copy_addr [take] [initialize]) of owned values to
/// - compose aggregates
/// - resolve phi interference
///
/// For guarantee values, this pass inserts neither copies nor
/// moves. Address-only values are potentially unmovable when borrowed. This
/// means that guaranteed address-only aggregates and phis are prohibited. This
/// SIL invariant is enforced by SILVerifier::checkOwnershipForwardingInst() and
/// SILVerifier::visitSILPhiArgument().
///
/// ## Step #1: Map opaque values
```
@rjmccall have I explained this well enough?

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -2,271 +2,521 @@
+/// not allowed.
+///
+/// When owned values are consumed by phis, multiple storage locations are
+/// required to avoid interfering with other phi operands. However, the value
```
In OSSA (and even SIL in general) it vastly simplifies everything to distinguish between phis and terminator results. The opcode names are complete nonsense, but in the SILArgument API, the PhiOperand abstraction, and all the code we've written in the past year, the distinction is very clear.

Added `(Phi values are block arguments in which phi's arguments are branch operands).`

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.h:
```diff
@@ -0,0 +1,282 @@
+/// directly reuse their operand's storage.
+///
+/// To materialize projections, address lowering follows the original def-use
+/// edges for address-only values. Consequently, values that have storage cannot
```
I'll try

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -2,271 +2,535 @@
+
+/// Invoke \p cleanup on all paths exiting a call.
+static void
+cleanupAfterCall(FullApplySite apply,
```
I can't because `insertAfterFullEvaluation` creates its own builder

atrick committed:
[SIL-opaque] Code review suggestions

Mostly documentation and typos.

atrick commented:
> I'll continue reviewing next week

It seems very time consuming to fill out a git comment for every typo or minor suggestion. I just copy-paste all the suggestions into one file while reviewing. No line numbers needed.

eeckstein commented:
> It seems very time consuming to fill out a git comment for every typo or minor suggestion. 

No, that's fine for me.

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -282,312 +546,656 @@ class OpaqueValueVisitor {
+      auto *openingInst = openingVal->getDefiningInstruction();
+      assert(openingVal && "all opened archetypes should be resolved");
+      if (firstOpeningInst
+          && pass.domInfo->dominates(firstOpeningInst, openingInst)) {
```
> So, it's not "as early as possible", but should be "as late as necessary".
> But IIUC, this code finds the earliest type-opening instruction.

You are correct that this is incorrect. I changed the code to find the "latestOpeningInstruction".

atrick committed:
Fix alloc_stack placement for open_existential.

Compute the latestOpeningInst, not the firstOpeningInst.

atrick committed:
Rename "phi copy" to "phi move" for consistency with documentation.

In classic compiler terminology, this is a "phi copy" algorithm. But
the documentation now tries to clearly distinguish between "semantics
copies" vs. moves, where moves are "storage copies".

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -282,312 +546,656 @@ class OpaqueValueVisitor {
+    }
+  });
+  auto allocPt = firstOpeningInst ? std::next(firstOpeningInst->getIterator())
+                                  : pass.function->begin()->begin();
```
@eeckstein Initially (4 years ago) I was paranoid about -Onone compile time. Not anymore. We certainly could do an analysis here to figure out the lifetime of stack slots.

I don't think shrink wrapping will make any difference until address lowering is run after inlining, so nothing needs to happen in this PR. It is, however, still worth thinking about.

We can't actually figure out the optimal alloc_stack placement until we know all the places where we'll end up storing into the reused location. This would be a completely separate analysis. I can't think of any reason to complicate address lowering with that analysis. If that's a valuable thing to do, then why should it be confined to this pass? And why confined to address-only values? Shouldn't we perform shrink wrapping on all stack allocations?

I'm open to all suggestions.

@aschwaighofer any ideas?

eeckstein replied:
My concern is that we have several peephole optimizations which only kick in if all uses of an `alloc_stack` are in the same block.

But I agree, this pass is already very complex. It would make more sense to implement stack-location shrink wrapping as a separate pass. But it might be necessary to do it before opaque values are turned on to not regress performance.

atrick replied:
> Another concern is that we have several peephole optimizations which only kick in if all uses of an alloc_stack are in the same block.

That's too bad. For the sake of optimizer stability, I don't think we should have any peephole optimizations that make such an assumption.

I agree that shrink-wrapping should be done before address lowering is enabled after Mem2Reg runs at -O. Can you think of any reason this would cause a problem before Mem2Reg?

Note that any address-only local variables will still have the same local alloc_stack scope prior to Mem2Reg.

I'm not sure we would want to do shrink-wrapping at -Onone. We would need to figure out how to preserve variable inspection after last-use.

eeckstein commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -282,312 +537,659 @@ class OpaqueValueVisitor {
+  // Instructions that produce an opened type never reach here because they
+  // have guaranteed ownership--they project their storage. We reach this
+  // point after the opened value has been copied.
+  if (auto *defInst = value->getDefiningInstruction()) {
```
This if-statement is not needed, you can directly do
```
if (auto *singleValue = dyn_cast<SingleValueInstruction>(value))
```
or, even better:
```
assert((!isa<SingleValueInstruction>(value) || !cast<SingleValueInstruction>(value)->getDefinedOpenedArchetype()) && "...")
```
Then you don't need the `#ifndef NDEBUG`

eeckstein commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -596,944 +1198,1991 @@ class AddressMaterialization {
+  SILValue materializeTupleExtract(SILInstruction *extractInst,
+                                   SILValue elementValue, unsigned fieldIdx);
+
+  SILValue materializeProjectionIntoUse(Operand *operand, bool intoPhiOperand);
```
`materializeProjectionIntoUse` sounds like the same as `materializeUseProjectionStorage`.
IIUC, `materializeUseProjectionStorage` adds some caching around `materializeProjectionIntoUse`.
Maybe you can find better names for these functions.

atrick replied:
> IIUC, materializeUseProjectionStorage adds some caching around materializeProjectionIntoUse

These are mutually recursive for materializing use projections. I renamed the more general entry point to `recursivelyMaterializeStorage`

eeckstein commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -596,944 +1198,1991 @@ class AddressMaterialization {
   SILValue def = operand->get();
-  SILValue destAddr;
-  if (operand->get()->getType().isAddressOnly(*pass.F)) {
+  if (def->getType().isAddressOnly(*pass.function)) {
```
Shouldn't this just be `def->getType().isAddress()`?
If it's an address type but not an address-only type we are creating a store here, which should be wrong.

atrick replied:
> Shouldn't this just be def->getType().isAddress()?

No. An address can't be passed into a struct/tuple. I renamed this to `initializeComposingUse`.

eeckstein commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -282,312 +537,659 @@ class OpaqueValueVisitor {
+
+void OpaqueStorageAllocation::allocatePhi(PhiValue phi) {
+  // Coalesces phi operand storage with the phi storage. The algorithm processes
+  // all incoming values at once, so it is is run when visiting the block
```
... it is run ...

eeckstein reviewed (commented):
next few comments

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.h:
```diff
@@ -0,0 +1,289 @@
+
+  /// For use-projections, identifies the operand index of the composing use.
+  /// Only valid for non-phi use projections.
+  uint16_t projectedOperandNum = InvalidOper;
```
> There are (crazy) situations where 16 bits are not enough. E.g. see https://github.com/apple/swift/pull/35289

@eeckstein In this case, I'll simply avoid projecting storage into a use beyond 64k operands. It is worthwhile for ValueStorage to be 2 words, and introducing a PointerIntPair is not worthwhile. Note that this actually still optimizes the preposterous test case you pointed out, because that type is nested.

In general, the way to handle the problem of importing a static array larger than 64k is definitely not to use 32-bits for an operand index. This will blow up the compiler in any many places where the complexity is cubic or worse when operands are assumed to scale proportionally with the function size.

The only hope of making the compiler handle types this large is to *actively* guard against analyzing operands lists larger then 64K. That means going through all the -Onone passes and inserting checks to bypass any analysis of those instructions.

atrick committed:
[SIL-opaque] avoid handling operands past 64k

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -282,312 +537,659 @@ class OpaqueValueVisitor {
+//===----------------------------------------------------------------------===//
+//                          OpaqueStorageAllocation
+//
+// Generate alloc_stack and address projections for abstract storage locations.
```
> address projections are actually created by AddressMaterialization and not here, right?

Yes

atrick committed:
[SIL-opaque] More file-level documentation

Explain high-level objectives and terminology with more precision.

atrick committed:
[SIL-opaque] Add section-level comment explaining storage allocation

atrick committed:
[SIL-opaque] remove cleanupAfterCall helper

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -2,271 +2,526 @@
+//                              ValueStorageMap
+//
+//              Map Opaque SILValues to abstract storage units.
+//===----------------------------------------------------------------------===//
```
This is now documented in two places but I think that's reasonable.

The def-projections are not "just optimizations". The existing file-level comment explains why...
https://github.com/apple/swift/pull/41557/commits/21326280bf3e6113f5f0ad878a95b8dd3f4a15bd

atrick committed:
[SIL-opaque] add a test case for phi coalescing

Avoid attempting to coalesce enum payloads.

atrick committed:
[SIL-opaque] in-depth top-level documentation for phi coalescing.

atrick commented on lib/SILOptimizer/Mandatory/AddressLowering.cpp:
```diff
@@ -282,312 +537,659 @@ class OpaqueValueVisitor {
+void OpaqueStorageAllocation::allocatePhi(PhiValue phi) {
+  // Coalesces phi operand storage with the phi storage. The algorithm processes
+  // all incoming values at once, so it is is run when visiting the block
+  // argument.
```
Yes, the PhiStorageOptimizer was not well documented at the top level. I'm pasting the file comment here because I can't refer to commits after rebasing
```
/// PhiStorageOptimizer implements an analysis used by AddressLowering
/// to reuse storage across block arguments.
///
/// In OSSA, phi operands can often be coalesced because they are
/// consuming--they end the lifetime of their operand. This optimization may
/// fail to coalesce an operand for two major reasons:
///
/// 1. This phi operand is already coalesced with other storage, possibly of a
/// different type:
///
///     %field = struct_extract %struct : $Struct<T>, #field
///     br bb(%field : $T)
///
///   bb(%phi : @owned $T):
///     ...
///
/// 2. This phi operand interferes with another coalesced phi operand.
///
/// Only one of the call results below, either %get0 or %get1, can be coalesced
/// with %phi. The %phi will itself be coalesced with this function's indirect
/// @out argument.
///
///   sil [ossa] @function : $@convention(thin) <T> () -> @out T {
///   bb0:
///     %get0 = apply %get<T>() : $@convention(thin) <τ_0_0>() -> @out τ_0_0
///     %get1 = apply %get<T>() : $@convention(thin) <τ_0_0>() -> @out τ_0_0
///     cond_br undef, bb2, bb1
///
///   bb1:
///     destroy_value %get0 : $T
///     br bb3(%get1 : $T)
///
///   bb2:
///     destroy_value %get1 : $T
///     br bb3(%get0 : $T)
///
///   bb3(%phi : @owned $T):
///     return %phi : $T
///
/// TODO: Liveness is currently recorded at the block level. This could be
/// extended to handle operand with nonoverlapping liveness in the same
/// block. In this case, %get0 and %get1 could both be coalesced with a bit of
/// extra book-keeping:
///
///   bb0:
///     %get0 = apply %get<T>() : $@convention(thin) <τ_0_0>() -> @out τ_0_0
///
///   bb1:
///     destroy_value %get0 : $T
///     %get1 = apply %get<T>() : $@convention(thin) <τ_0_0>() -> @out τ_0_0
///     br bb3(%get1 : $T)
///
///   bb2:
///     br bb3(%get0 : $T)
///
///   bb3(%phi : @owned $T):
```

atrick committed:
[SIL-opaque] minor NFC review feedback

atrick committed:
[SIL-opaque] rename materialization functions.

atrick committed:
[SIL-opaque] rename initializeComposingUse

atrick commented:
I'm planning to merge this PR so we can collaborate on bug fixes, particularly in SILGen. The pass is guarded by a flag, so it should have no effect on 5.7 functionality. We only need it in tree for testing other areas of the compiler.

atrick commented:
@swift-ci test
